core/src/test/java/org/elasticsearch/index/replication/IndexLevelReplicationTests.java

```diff
@@ -28,4 +40,64 @@ public void testSimpleReplication() throws Exception {
             shards.assertAllEqual(docCount);
         }
     }
+
+    public void testSimpleAppendOnlyReplication() throws Exception {
+        try (ReplicationGroup shards = createGroup(randomInt(2))) {
+            shards.startAll();
+            final int docCount = randomInt(50);
+            shards.appendDocs(docCount);
+            shards.assertAllEqual(docCount);
+        }
+    }
+
+    public void testAppendWhileRecovering() throws Exception {
+        try (ReplicationGroup shards = createGroup(0)) {
+            shards.startAll();
+            IndexShard replica = shards.addReplica();
+            CountDownLatch latch = new CountDownLatch(2);
+            int numDocs = randomIntBetween(100, 200);
+            shards.appendDocs(1);// just append one to the translog so we can assert below
+            Thread thread = new Thread() {
+                @Override
+                public void run() {
+                    try {
+                        latch.countDown();
+                        latch.await();
+                        shards.appendDocs(numDocs-1);
+                    } catch (Exception e) {
+                        throw new AssertionError(e);
+                    }
+                }
+            };
+            thread.start();
+            Future<Void> future = shards.asyncRecoverReplica(replica, (indexShard, node)
+                -> new RecoveryTarget(indexShard, node, recoveryListener, version -> {}) {
+                @Override
+                public void cleanFiles(int totalTranslogOps, Store.MetadataSnapshot sourceMetaData) throws IOException {
+                    super.cleanFiles(totalTranslogOps, sourceMetaData);
+                    latch.countDown();
+                    try {
+                        latch.await();
```

Comment:
I wonder we should remove the concurrency here and fix the time the document is indexed to this moment - either by using two latches or just calling `shards.appendDocs` here?